Got Docker-compose working again

Fokko opened the pull request:
Hi Guys,

First of all, thanks for sharing the HelloLily CRM, loving it.

I had some troubles with the docker-compose configuration. I've removed some stuff get it back to work again. I've also did some optimizations of merging `RUN` commands to get more optimized layers without unused files. Also I've created `start.sh` to rebuild the Docker stack.

Cheers, Fokko.

epasagic commented:
Hi Fokko,

So awesome that you've contributed to the project. Also makes you the first non-Spindle contributor ;) We'll test these changes and get back to you ASAP. For now the only comment I have if the fact that the develop branch has been merged into your branch too. Maybe something went wrong with rebasing?

Regards,

Ednan

Fokko commented:
Hi Ednan,

I see, let me work on that. There is also a small issue with the UI, I still get 404's on all the css/js. Will fix this in the next commit, please wait with the merge.

Cheers, Fokko.

ferdynice commented:
Hi, thank you for providing this pull request. However, it seems a bit to generic to merge back. There are various changes, some of it may be useful. Some changes are not needed in my opinion, such as the changes to the .dockerignore and also the start script. The start script is very much tailored to a specific workflow. With the right aliases, I think the docker-compose commands are flexible enough to use on their own, so we do not need to impose a workflow on developers.

My suggestion, be specific and concise in your pull request in what you want to see changed, so we can discuss it properly and also merge it back when it is suitable.

In any case, thanks for trying out HelloLily.

snoepkast commented on Dockerfile:
```diff
@@ -1,54 +1,46 @@
+RUN mkdir -p /root/.config/pudb && pip install pudb
+
+WORKDIR /root/
+ADD . /root/
```
The add command doesn't copy the current dir name so
```
ADD . /root/lily/
WORKDIR /root/lily/
```
Would be better I think.

ferdynice replied:
Please note that we do not use the workflow to copy all files to the image, because this is not very much suited for developing. Instead we mount the source files needed for running the project, so changes in code are directly reflected on the local development server.

snoepkast replied:
Ah I didn't realize that, that would also explain his need for the `start.sh` which we don't have.

snoepkast commented:
I don't completely agree with @ferdynice, the only thing I do agree with is the `start.sh` file. There are a lot of good things in this pull request, and I would like to see it merged (eventually).

However for the future, please be precise in reporting the error you had (and fixed) so we can better check the pull request. Also seperate different functionalities in different pull requests (Bugfix vs optimization vs workflow change).

Fokko commented:
Hi Guys,

Thanks for the response. Please allow me to elaborate a bit:

- I've updated some of the Dockerfiles based on the [best practices](https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/) supplied by Docker. For example, using `apt-get clean` to remove temporary files to avoid leaving unused files behind in the layers. Downloading a file, and removing it in a different `RUN` wil cause the file to be between the layers and thus increasing the size of the image. Therefore I also merged a couple of `RUN` commands. When the application is getting to a stable version, it would also be nice to remove for example `gulf` and `npm` after using it, to decrease size of the image.
- The `oauth2client.django_orm` is deprecated, and updated to `oauth2client.contrib.django_orm`.
- I have removed the start script. Personally I like it to rebuild a clean set of images/containers after changes to the code/files.
- Linking volumes does not work when the images is submitted to the [docker-hub](https://docs.docker.com/docker-hub/builds/), which would be very nice. Then the `docker-compose` can make use of the pre-build images (which are build every time there is a push to a specific branch), which is very nice for people who just want to give HelloLily a try.

For linking the directory as a volume makes sense for development. I suggest using a different `docker-compose -f docker-comse-development.yml`. Please keep in mind that linking is on the table for deprecation in Docker (relying on docker networks and volumes).

Let me know what you think.

Cheers, Fokko.

ferdynice commented:
I think the `apt-get clean && ...` is a great idea, to be complete you can change it to `apt-get clean && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*`.

ferdynice commented:
About the adding source to the image and pushing to docker-hub: In this stage of the project we like to focus the docker-compose.yml on developing. To run the project, you still need to have docker-compose for the various services. While I definitely see a future where we bake HelloLily images with sources, so you can run the project more easily, I don't believe this should now get in the way of developing.

ferdynice commented:
The "oauth2client.django_orm" change is fine. Preferably this goes in a separate pr, but for now it's ok.

Fokko commented:
Nice, good suggestion of the extended `rm`, changed it and everything still works fine. I also restored the original compose file as the default so you guys don't have to change anything.

Cheers, Fokko.

ferdynice commented:
Sure. Don't forget to restore the .dockerignore and Dockerfile (except of course the optimization changes).

ferdynice commented:
We now actually ran into the error you fixed with oauth and we are keen to committing this to develop. Because we do not want to commit the entire PR, please let us know if you wish to change the PR to the set of changes we discussed above. Otherwise we will cherrypick the changes ourself so we can move on. (And close / discard the PR).

In any case, thanks again for contributing!
